flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.flink.api.common.operators.base.GroupReduceOperatorBase;
+import org.apache.flink.api.common.operators.base.MapOperatorBase;
+import org.apache.flink.api.common.operators.base.MapPartitionOperatorBase;
+import org.apache.flink.api.common.operators.util.FieldList;
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeComparatorFactory;
+import org.apache.flink.api.java.typeutils.RecordTypeInfo;
+import org.apache.flink.runtime.io.network.DataExchangeMode;
+import org.apache.flink.runtime.operators.udf.AssignRangeIndex;
+import org.apache.flink.runtime.operators.udf.PartitionIDRemoveWrapper;
+import org.apache.flink.runtime.operators.udf.RangeBoundaryBuilder;
+import org.apache.flink.api.java.functions.SampleInCoordinator;
+import org.apache.flink.api.java.functions.SampleInPartition;
+import org.apache.flink.api.java.sampling.IntermediateSampleData;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+import org.apache.flink.api.java.typeutils.TypeExtractor;
+import org.apache.flink.optimizer.dag.GroupReduceNode;
+import org.apache.flink.optimizer.dag.MapNode;
+import org.apache.flink.optimizer.dag.MapPartitionNode;
+import org.apache.flink.optimizer.dag.TempMode;
+import org.apache.flink.optimizer.plan.Channel;
+import org.apache.flink.optimizer.plan.NamedChannel;
+import org.apache.flink.optimizer.plan.OptimizedPlan;
+import org.apache.flink.optimizer.plan.PlanNode;
+import org.apache.flink.optimizer.plan.SingleInputPlanNode;
+import org.apache.flink.optimizer.util.Utils;
+import org.apache.flink.runtime.operators.DriverStrategy;
+import org.apache.flink.runtime.operators.shipping.ShipStrategyType;
+import org.apache.flink.util.Visitor;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+public class RangePartitionRewriter implements Visitor<PlanNode> {
+	final static long SEED = org.apache.flink.api.java.Utils.RNG.nextLong();
+
+	final OptimizedPlan plan;
+
+	public RangePartitionRewriter(OptimizedPlan plan) {
+		this.plan = plan;
+	}
+
+	@Override
+	public boolean preVisit(PlanNode visitable) {
+		return true;
+	}
+
+	@Override
+	public void postVisit(PlanNode visitable) {
+		final List<Channel> outgoingChannels = visitable.getOutgoingChannels();
+		final List<Channel> newOutGoingChannels = new LinkedList<>();
+		final List<Channel> toBeRemoveChannels = new ArrayList<>();
+		for (Channel channel : outgoingChannels) {
+			ShipStrategyType shipStrategy = channel.getShipStrategy();
+			if (shipStrategy == ShipStrategyType.PARTITION_RANGE) {
+				TypeInformation<?> outputType = channel.getSource().getProgramOperator().getOperatorInfo().getOutputType();
+				// Do not optimize for record type, it's a special case for range partitioner, and should be removed later.
+				if (!(outputType instanceof RecordTypeInfo)) {
```

Comment:
This will continue in case of a Record type and cause silently wrong results, no? Can we do this check earlier, e.g., in the `PartitionOperator` and throw an `UnsupportedOperationException`? We should also add a note to FLINK-1278 in order to remember to remove the exception later when FLINK-1278 is resolved.